### PR TITLE
fix(experience): update hooks

### PR DESCRIPTION
**Pull request by smalexan**
## Description of this change

Register willShowExperience and experienceHidden
Call willShowExperience and experience at appropriate times 

## Why is this change being made?
- [x] Bug fix (non-breaking change that fixes an issue)
- [ ] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist

- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed.
- [x] I have evaluated the security impact of this change, and [OWASP Secure Coding Practices](https://owasp.org/www-project-secure-coding-practices-quick-reference-guide/migrated_content#) have been observed.
- [x] I have informed stakeholders of my changes.

### CODE REVIEW

File: src/pure.ts [hunk 1]
```diff
@@ -454,7 +471,7 @@ export class Ketch {
 
             // experience will not show - call functions registered using onHideExperience
             this._hideExperience.forEach(func => {
-              func();
+              func("willNotShow");
```
Comment (smalexan) [inline src/pure.ts:474]: Should this be a constant? Or just live in documentation? Recommendations on tracking something like this?

File: src/index.ts [hunk 2]
```diff
@@ -85,7 +85,7 @@ function getAction(action: string): Function | undefined {
     case 'onEnvironment': return ketch.onEnvironment;
     case 'onGeoIP': return ketch.onGeoIP;
     case 'onHideExperience': return ketch.onHideExperience;
-    case 'onShowExperience': return ketch.onShowExperience;
+    case 'onWillShowExperience': return ketch.onWillShowExperience;
```
Comment (smalexan) [inline src/index.ts:88]: This is not currently used by customers and callbacks were never previously called in code

File: src/pure.ts [hunk 3]
```diff
@@ -1197,7 +1214,7 @@ export class Ketch {
 
     // Call functions registered using onHideExperience
     this._hideExperience.forEach(function (func) {
-      func();
+      func(reason);
```
Comment (smalexan) [inline src/pure.ts:1217]: reasons lanyard may close are `setConsent`, `invokeRight`, or `close`. Other possible value defined by ketch-tag is `willNotShow`. Recommendations on how to best handshake?

File: src/pure.ts (hunk 1 above)
Comment (sethyates) [reply src/pure.ts:474]: huh? I don't even understand what this is doing.

File: src/pure.ts (hunk 3 above)
Comment (sethyates) [reply src/pure.ts:1217]: Enum

File: src/pure.ts (hunk 1 above)
Comment (smalexan) [reply src/pure.ts:474]: Want to be able to notify plugins / site that no experience will show thus they can render any popups and/or execute some functionality. This is likely similar functionality to what will execute after an experience is shown and then closes

File: src/pure.ts (hunk 3 above)
Comment (smalexan) [reply src/pure.ts:1239]: Done

**Review by sethyates — APPROVED** [review]

**Pull request merged**